staker/staker.gno

```diff
@@ -0,0 +1,248 @@
+package staker
+
+import (
+	"errors"
+	"std"
+	"strconv"
+	"time"
+
+	"gno.land/p/demo/grc/grc20"
+	u256 "gno.land/p/demo/uint256"
+)
+
+var (
+	ErrInsufficientBalance = errors.New("insufficient balance")
+	ErrUnauthorized        = errors.New("unauthorized")
+	ErrInvalidTimeRange    = errors.New("invalid time range")
+	ErrNoData              = errors.New("no delegation data for the given period")
+)
+
+// DelegationHistory represents a single delegation event
+type DelegationHistory struct {
+	amount    *u256.Uint
+	timestamp uint64
+}
+
+// staking contract
+type Staker struct {
+	token             grc20.IGRC20
+	delegations       map[std.Address]*u256.Uint
+	delegationHistory map[std.Address][]DelegationHistory
+	totalStaked       *u256.Uint
+}
+
+func New(token grc20.IGRC20) *Staker {
+	return &Staker{
+		token:             token,
+		delegations:       make(map[std.Address][]*u256.Uint),
+		delegationHistory: make(map[std.Address][]DelegationHistory),
+		totalStaked:       u256.NewUint(0),
+	}
+}
+
+// StakeAmount stakes the specified amount for the given delegate
+func (s *Staker) StakeAmount(delegate std.Address, amount *u256.Uint) error {
+	caller := std.GetOrigCaller()
+
+	// Check if the caller has enough balance
+	balance, err := s.token.BalanceOf(caller)
+	if err != nil {
+		return err
```

Comment:
most examples in gno repo `panic` rather than returning an `error` (same goes for gnoswap).

@mconcat WDYT?